controller/apps.go

```diff
@@ -0,0 +1,413 @@
+	}
+	return oc, nil
+}
+
+func (c *AppsController) getSpaceNameFromSpaceID(ctx context.Context, spaceID uuid.UUID) (*string, error) {
+	// use WIT API to convert Space UUID to Space name
+	oc, err := c.getAndCheckOsioClient(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	osioSpace, err := oc.GetSpaceByID(spaceID.String(), false)
+	if err != nil {
+		return nil, errors.NewNotFoundError("osio space", spaceID.String())
+	}
+	return osioSpace.Attributes.Name, nil
+}
+
+func (c *AppsController) getNamespaceName(ctx context.Context) (*string, error) {
+	osioclient, err := c.getAndCheckOsioClient(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	kubeSpaceAttr, err := osioclient.GetNamespaceByType(nil, "user")
+	if err != nil {
+		return nil, err
+	}
+	if kubeSpaceAttr == nil || kubeSpaceAttr.Name == nil {
+		return nil, errors.NewNotFoundError("namespace", "user")
+	}
+
+	return kubeSpaceAttr.Name, nil
+}
+
+func (c *AppsController) getKubeClient(ctx context.Context) (*KubeClient, error) {
+
+	// create Auth API login object
+	authClient, err := NewAuthClient(ctx, c.AuthURL)
+	if err != nil {
+		goa.LogInfo(ctx, "errror creating auth client:"+tostring(err))
+		return nil, err
+	}
+
+	// get the user definition (for cluster URL)
+	authUser, err := authClient.getAuthUser()
+	if err != nil {
+		goa.LogInfo(ctx, "error accessing Auth server"+tostring(err))
+		return nil, err
+	}
+
+	if authUser == nil || authUser.Data.Attributes.Cluster == nil {
+		goa.LogInfo(ctx, "error getting user from Auth server:"+tostring(authUser))
+		return nil, nil
+	}
+
+	// get the login token for the cluster OpenShift API
+	osauth, err := authClient.getAuthToken(*authUser.Data.Attributes.Cluster)
+	if err != nil {
+		goa.LogInfo(ctx, "error getting openshift credentials:"+tostring(err))
+		return nil, err
+	}
+
+	kubeURL := *authUser.Data.Attributes.Cluster
+	kubeToken := *osauth.AccessToken
+
+	kubeNamespaceName, err := c.getNamespaceName(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// create the cluster login object
+	kc, err := NewKubeClient(kubeURL, kubeToken, *kubeNamespaceName)
+	if err != nil {
+		return nil, err
+	}
+	return kc, nil
+}
+
+func (c *AppsController) getAndCheckKubeClient(ctx context.Context) (*KubeClient, error) {
+
+	kc, err := c.getKubeClient(ctx)
+	if err != nil {
+		goa.LogInfo(ctx, "didn't actually get a token")
+		return nil, errors.NewUnauthorizedError("openshift token")
+	}
+	return kc, nil
+}
+
+// SetDeployment runs the setDeployment action.
+func (c *AppsController) SetDeployment(ctx *app.SetDeploymentAppsContext) error {
+	// AppsController_SetDeployment: start_implement
+
+	if ctx.PodCount == nil {
+		// TODO this should be error 400 (bad request) not 404 (not found)
+		return errors.NewNotFoundError("parameter", "podCount")
+	}
+
+	kc, err := c.getAndCheckKubeClient(ctx)
+	if err != nil {
+		return err
+	}
+
+	kubeSpaceName, err := c.getSpaceNameFromSpaceID(ctx, ctx.SpaceID)
+	if err != nil {
+		return err
```

Comment:
you should make sure that the method returns a `400 Bad Parameter` or `404 Not Found` error. If it's just an `error`, then you should convert it here. I personally prefer the conversion just before returning from the controller, not deeper in the code.